File: typescript/merge_sort.ts
import mergeSortedArrays from "./merge_sorted_arrays";

/**
 * Merge sort algorithm.
 * @param input - Array of numbers to sort.
 * @returns Sorted array of numbers.
 */
function mergeSort(input: number[]): number[] {
  if (input.length <= 1) {
    return input;
  }
  const middle = Math.floor(input.length / 2);
  const left = input.slice(0, middle);
  const right = input.slice(middle);

  return mergeSortedArrays(mergeSort(left), mergeSort(right));
}

export default mergeSort;

File: csharp/MergeSort.cs
/// <summary>
/// Merge sort algorithm.
/// </summary>
/// <param name="input">Array of integers to sort.</param>
/// <returns>Sorted array of integers.</returns>
private int[] MergeSort(int[] input)
{
  if (input.Length <= 1)
  {
    return input;
  }

  int middle = (int)Math.Floor((double)input.Length / 2);
  int[] left = input.Take(middle).ToArray();
  int[] right = input.Skip(middle).ToArray();

  // see "./MergeSortedArrays.cs" implementation of "MergeSortedArrays"
  return MergeSortedArrays(MergeSort(left), MergeSort(right));
}